Jewish calendar config flow

tsvi opened the pull request:
## Breaking Change:

<!-- What is breaking and why we have to break it. Remove this section only if it was NOT a breaking change. -->

## Description:

Moving the Jewish calendar to config flow and adding support for multiple entries (this might be interesting to support times for different locations)

**Related issue (if applicable):** fixes #<home-assistant issue number goes here>

**Pull request with documentation for [home-assistant.io](https://github.com/home-assistant/home-assistant.io) (if applicable):** home-assistant/home-assistant.io#11176

## Example entry for `configuration.yaml` (if applicable):
```yaml

```

## Checklist:
  - [x] The code change is tested and works locally.
  - [x] Local tests pass with `tox`. **Your PR cannot be merged unless tests pass**
  - [x] There is no commented out code in this PR.
  - [x] I have followed the [development checklist][dev-checklist]

If user exposed functionality or configuration variables are added/changed:
  - [x] Documentation added/updated in [home-assistant.io](https://github.com/home-assistant/home-assistant.io)

If the code communicates with devices, web services, or third-party tools:
  - [x] [_The manifest file_][manifest-docs] has all fields filled out correctly. Update and include derived files by running `python3 -m script.hassfest`.
  - [x] New or updated dependencies have been added to `requirements_all.txt` by running `python3 -m script.gen_requirements_all`.
  - [x] Untested files have been added to `.coveragerc`.

If the code does not interact with devices:
  - [x] Tests have been added to verify that the new code works.

[dev-checklist]: https://developers.home-assistant.io/docs/en/development_checklist.html
[manifest-docs]: https://developers.home-assistant.io/docs/en/creating_integration_manifest.html

tsvi committed:
Use scaffold to create config flow for jewish calendar

tsvi committed:
Cleanup config flow

balloob commented on homeassistant/components/jewish_calendar/config_flow.py:
```diff
@@ -0,0 +1,56 @@
+                        ),
+                    },
+                )
+            except Exception:  # pylint: disable=broad-except
```
Nothing should be able to raise here, so you can remove this ?

tsvi replied:
You're right. Fixed

balloob commented on homeassistant/components/jewish_calendar/__init__.py:
```diff
@@ -39,49 +44,25 @@
     """Set up the Jewish Calendar component."""
-    name = config[DOMAIN][CONF_NAME]
-    language = config[DOMAIN][CONF_LANGUAGE]
+    return True
```
You need to store the configuration that is part of `config[DOMAIN]` if you want to keep supporting configuration.yaml. If you don't, update `CONFIG_SCHEMA` to only optionally accept an empty dict

tsvi replied:
Copied from the abode integration adding support for both config_entries and configuration.yaml
Still failing tests. I suppose that fixing this should obsolete your comment on the tests?

balloob commented:
Your tests are failing because your tests rely on setting it up via configuration.yaml, but you have removed that support. 

What you can do is to add a mock config entry before setting up the integration, and the config entry will be set up. 

```python
from tests.common import MockConfigEntry

MockConfigEntry(domain="jewish_calendar", data={"your_data": "bla"})

assert await async_setup_component(hass, "jewish_calendar", {})
```

tsvi committed:
Remove unreachable exception

tsvi committed:
Support configuring via configuration.yaml as well

Code copied from the abode integration

tsvi commented:
Ok, so looking more deeply at the code it seems I need to implement `async_step_import`. TBD tomorrow.

tsvi committed:
Fix sensor and binary sensor tests

tsvi committed:
Fix strings.json

tsvi committed:
Fix convert schema error

tsvi committed:
Make config flow test pass

tsvi committed:
Add translations and make form show up

tsvi committed:
Rewrite strings so they look nicer

tsvi committed:
Make coordinates optional

When set to Inclusive, the UI won't allow you to skip them.
Also reorder the settings logically

tsvi commented:
TODO:
 - Add test for import
 - Check possibility of editing the settings at a later time
 - See if we can have another way of setting vol.Inclusive and it being optional

tsvi commented:
@frenck Do I need updated docs for adding config_flow support?  If so, do you have an example where it is done nicely?

tsvi committed:
Add tests for import step

tsvi committed:
Make pylint happy ->  remove DOMAIN

tsvi committed:
Revert "Make pylint happy ->  remove DOMAIN"

This reverts commit 75345606c496b4cdbe491976ab8b102a0af6318d.

tsvi committed:
Use asyncio.gather for unloading entries

tsvi committed:
Remove period at the end of logging statement

tsvi commented on homeassistant/components/jewish_calendar/__init__.py:
```diff
@@ -99,10 +93,18 @@
+async def async_unload_entry(hass, entry):
+    """Unload a config entry."""
+    await hass.config_entries.async_forward_entry_unload(entry, "sensor")
+    await hass.config_entries.async_forward_entry_unload(entry, "binary_sensor")
```
Fixed

tsvi commented on homeassistant/components/jewish_calendar/config_flow.py:
```diff
@@ -0,0 +1,65 @@
+                    CONF_NAME: user_input.get(CONF_NAME, DEFAULT_NAME),
+                    CONF_DIASPORA: user_input.get(CONF_DIASPORA, DEFAULT_DIASPORA),
+                    CONF_LANGUAGE: user_input.get(CONF_LANGUAGE, DEFAULT_LANGUAGE),
+                    CONF_CANDLE_LIGHT_MINUTES: user_input.get(
```
Ok, only diaspora and language are critical.

I'll modify to use a options flow.

tsvi commented on homeassistant/components/jewish_calendar/config_flow.py:
```diff
@@ -0,0 +1,65 @@
+    async def async_step_import(self, import_config):
+        """Import a config entry from configuration.yaml."""
+        if self._async_current_entries():
+            _LOGGER.warning("Only one configuration of Jewish calendar is allowed.")
```
Fixed

tsvi committed:
Implement options flow and add support for multiple entries

tsvi committed:
Fixes due to multiple entries

tsvi committed:
Update strings.json to reflect changes to options flow

MartinHjelmare commented on homeassistant/components/jewish_calendar/config_flow.py:
```diff
@@ -28,32 +29,40 @@ class JewishCalendarConfigFlow(config_entries.ConfigFlow, domain=DOMAIN):
+                    ),
+                }
+            ),
+            errors={},
```
We should show the stored errors here.

tsvi replied:
Right, missed that. I used transmission client as a template and forgot I brought back the errors.

tsvi committed:
Pass the errors to async_show_form

tsvi committed:
Update tests to reflect on use of options flow

tsvi committed:
Fix-up en.json

tsvi committed:
Fix strings.json

tsvi committed:
Fix test_config_flow.py

tsvi committed:
Fix docstrings

tsvi committed:
No need to pop items from dict

Get is more than enough.

tsvi committed:
Cleanup location information needed by Jewish Calendar

tsvi committed:
Make flake8 happy: remove unused imports

tsvi commented:
Closing pull request temporarily as I want to split some of the changes into multiple PR
